### ISSUE TITLE
How to render form layout, in both form and form_child scenario

### ISSUE THREAD

**Issue by rudiedirkx** [kristijanhusak/laravel-form-builder#244]
I have an `OrganizationForm` that is used in 2 forms:

* creating an organization has an 'organization' fieldset with loads the `OrganizationForm` (with `ChildFormType`)
* editing an organization IS the `OrganizationForm`

It has a pretty sophisticated layout of fields, with internal fieldsets (only markup). I want to reuse that layout for both implementations: as a full form and as a child form. Is there a way?

I think what I need is the fields part of templates `form.php` and `child_form.php` to be the same, and configurable the same.

I can completely override `form.php` and `child_form.php` and maybe/almost share a `fields.php`, but the fields itself should be enough. Is there a good DRY way to only override the fields layout, for both the `form` and the `child_form` scenario?

**Comment by kristijanhusak** [comment]
If you can post some code, it's easier for me to understand it that way.

**Comment by rudiedirkx** [comment]
Alright. I'll try to be succinct.

I have 2 page forms:

* `OrganizationForm` contains the organization fields (there's a matching model `Organization`)
* `CreateOrganizationForm` contains 2 child forms: `OrganizationForm` and `UserForm`, because when you create a new organization, you also create an organization admin (user)

In both forms, the organization has the same many fields, in several fieldsets (just markup/layout). I'd like to share that field layout logic between both forms. I was hoping I could tell `OrganizationForm` to always render itself like `organization.partials.form`, but that apparently doesn't work for child forms, because they render as a field, not as a form.

This is the code I have now, to share the field layout logic between master form and child form:

`CreateOrganizationForm` includes both forms, and how to render the Organization part:

```
class CreateOrganizationForm extends Form {
  public function buildForm() {
    $this->add('organization', 'form', [
      'class' => OrganizationForm::class,
      'template' => 'organization.partials.child_form', // This is the child form, without <form> tags
    ]);

    $this->add('admin', 'form', [
      'class' => UserForm::class,
    ]);
  }
}
```

`organization/partials/child_form.twig` extends the normal `child_form` tpl (in Twig, sorry):

```
{% if showLabel and showField and options.wrapperAttrs %}
  <fieldset {{ options.wrapperAttrs|raw }}>
{% endif %}

{% if showLabel and options.label %}
  <legend {{ options.labelAttrs|raw }}>{{ options.label }}</legend>
{% endif %}

{% if showField %}
  {% include 'organization.partials.fields' with {'fields': options.children} %}
{% endif %}

{% if showLabel and showField and options.wrapperAttrs %}
  </fieldset>
{% endif %}
```

which includes `organization/partials/fields.twig` with the big form field layout and more sub fieldsets. (It's a big form.) There are no `<form>` tags in there, because it's a known child form.

`OrganizationForm` contains all the table fields (and some more) in a flat hierarchy:

```
class OrganizationForm extends Form {
  public function buildForm() {
    $this->setFormOption('template', 'organization.partials.form'); // This is the master form, including <form> tags

    $this->add('name', 'text', [
      'label' => trans('organization.name'),
      'rules' => ['required', 'min:2'],
    ]);

    // .. millions of more fields here
  }
}
```

And finally there's `organization/partials/form.twig` which also includes the fields partial:

```
{% if showStart %}
  {{ form_open(formOptions) }}
{% endif %}

{% if form.model and form.model.logo_path %}
  <img src="{{ form.model.logo_path }}" alt="{{ trans('organization.logo') }}" />
{% endif %}

{% if showFields %}
  {% include 'organization.partials.fields' with {'fields': form} %}
{% endif %}

{% if showEnd %}
  {{ form_group(form, ['submit']) }}

  {{ form_close(formOptions) }}
{% endif %}
```

The fields partial is big, but there's a bunch of these:

```
<fieldset class="organization-address">
  <legend>{{ trans('organization.organization_details') }}</legend>
  {{ form_group(fields, ['name', 'organization_address', 'website', 'client_since', 'client_until'], {'wrapper': {'class': 'form-inline'}}) }}
</fieldset>
```

The fields partial takes all fields from OrganizationForm and marks them up prettily. The exact same kind of pretty when a form on its own (`OrganizationForm`) or a child form of another (`CreateOrganizationForm`).

The current setup is doable, but there's a lot of overriding with mostly no change at all. The only change from the original (your) templates, is the fields part.

What I had hoped to do was specify the 'field template' once, and be done with it. Something like this:

```
class OrganizationForm extends Form {
  public function buildForm() {
    $this->setFormOption('fields_template', 'organization.partials.fields'); // Loads the field template wherever this form is
```

I guess the/my problem is in this line, which doesn't do all I had hoped:

```
$this->setFormOption('template', 'organization.partials.form'); // This is the master form, including <form> tags
```
That template is only used when the form is rendered as a full form, not a child form. But I only care about the field layout, not the `<form>` tags or submit button.

I could make a patch that defines the `fields_template` option, but I'm not sure it would be backwards compatible...

I'm afraid that wasn't very succinct...

**Comment by rudiedirkx** [comment]
There's another problem that made sharing the fields partial difficult, but #245 might be the answer to that:

The fields template uses `form_group($form, [$field_name, $field_name])` to render fields, but `$form` is different per scenario:

* The master form tpl has a real Form, with `getField()` etc
* The child form (field) only has an array of fields, not a `getField()` method, so the interface is different

Maybe `Form` and `ChildFormType` can have the same (limited) interface, so `getField()` would work on both, and I don't have to `with {'fields': options.children}` in the include. I couldn't find the ChildFormType object (`this`?), so I used the array vs Form =(

**Comment by rudiedirkx** [comment]
Thoughts? Sorry for the long story.

I think we should do 2 things:

1. Expose the `ChildFormType` object to `child_form.php` tpl as `$form`, so it can be used as a normal form. It's completely inaccessible now, or at least I couldn't find it, which is annoying.
2. Create a new `fields.php` tpl that's reused for both scenario's renders.

I think 1 is very easy. There might be some type hinting conflicts in form helpers?, since `Form` isn't a `ChildFormType` or vice versa.

I'll try a PR for 1 with a fresh Laravel install.

**Comment by rudiedirkx** [comment]
PR for option 1: https://github.com/kristijanhusak/laravel-form-builder/pull/256

**Comment by kristijanhusak** [comment]
Closing this since it's done here https://github.com/kristijanhusak/laravel-form-builder/pull/262 and merged to version 1.8.0

**Comment by rudiedirkx** [comment]
Weeelll, it's not really done. I'd like to keep this open until I've tried a `fields.php` tpl that works automagically. Will be a separate PR, but this is a todo reminder.
